text/unify-theorem-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,193 @@
+           | : @type
+```
+
+with the following newly supported attributes:
+
+- `recursive` and `corecursive` to declare a fixpoint / cofixpoint:
+
+  - attribute `recursive` is implicit for `Fixpoint` and rejected by
+    `CoFixpoint`;
+  - attribute `corecursive` is implicit for `CoFixpoint` and rejected
+    by `Fixpoint`;
+  - attribute `recursive` or `corecursive` is compulsory to use the
+    `with` variant of `Definition`;
+  - using the `with` variant of `Theorem` without attribute
+    `recursive` or `corecursive` raises a deprecation message and is
+    interpreted as if `recursive` had been used;
+  - the `@fixannot` clause is reserved to the `recursive` attribute
+    being used (or implicit in the case of `Fixpoint`).
+
+- `opaque` and `transparent` to declare an object opaque or
+  transparent like `Qed` and `Defined` do (note that this is not the
+  same as what the `Opaque` and `Transparent` commands do):
+
+  - `transparent` is assumed by default for `Definition`, `Fixpoint`,
+    `CoFixpoint` but `opaque` is accepted;
+  - `opaque` is assumed by default for `Theorem` and variants but
+    `transparent` is accepted.
+  - these attributes are only accepted when a body is provided (when
+    no body is provided the transparency is determined by the ending
+    `Defined` or `Qed`).
+
+# Drawbacks
+
+This creates more ways of doing the same thing.  People could decide
+to write things like:
+
+```coq
+#[ transparent ]
+Theorem not (b : bool) :=
+  match b with true => false | false => true end.
+
+#[ opaque ]
+Definition not_not_idem (b : bool) : not (not b) = b :=
+  match b with true => eq_refl | false => eq_refl end.
+```
+
+But it seems consistent with the current philosophy of the language to
+give many synonyms for doing the same thing and trusting the user to
+produce readable scripts in the end.
+
+# Alternatives
+
+We could keep the syntax separate and keep waiting for users asking
+for this or that extension to implement it, but it is unlikely that
+the syntax would ever converge, and it would remain inconsistent and
+confusing.  The first chapter of the refactored reference manual could
+rely on the unsatisfactory `Proof @term` syntax to avoid bringing the
+tactics in.
+
+We could decide to go with this plan but not provide the `transparent`
+and `opaque` attributes.  This means that when a body is provided,
+`Theorem` and `Definition` would not be interchangeable (but they
+already mostly are when in proof mode because the `Qed` and `Defined`
+commands are not forced upon the user).
+
+We could also decide to implement this plan but make `Definition` and
+`Theorem` fully interchangeable.  A `Theorem` with a body would be
+transparent by default, an `opaque` attribute could be used on
+`Definition` or `Theorem` to make the body opaque.
+
+# Unresolved questions
+
+The name of attributes `opaque` and `transparent` may be confusing
+with the `Opaque` and `Transparent` commands which are not as strong.
```

Comment:
Let me repeat again my call from six years ago to have more notions of opacity: https://github.com/coq/coq/issues/3389
This is probably beyond the scope of this PR

Reply:
That's interesting and also echoes the comment of Enrico above. But yes, this is out of scope.